src/content/contribute/writing-a-loader.md

```diff
@@ -48,93 +39,111 @@ resolveLoader: {
+
+Take the case of rendering a template file with data specified via loader options or query parameters. It could be written as a single loader that compiles the template from source, executes it and returns a module that exports a string containing the HTML code. However, in accordance with guidelines, a simple `apply-loader` exists that can be chained with other open source loaders:
 
-### Do not keep state between runs and modules
+- `jade-loader`: Convert template to a module that exports a function.
+- `apply-loader`: Executes the function with loader options and returns raw HTML.
+- `html-loader`: Accepts HTML and outputs a valid JavaScript module.
 
-A loader should be independent of other modules compiled (except of these issued by the loader).
+T> The fact that loaders can be chained also means they don't necessarily have to output JavaScript. As long as the next loader in the chain can handle its output, the loader can return any type of module.
 
-A loader should be independent of previous compilations of the same module.
+### Modular
 
-### Use the [loader-utils](https://github.com/webpack/loader-utils)
+Keep the output modular. Loader generated modules should respect the same design principles as normal modules.
 
-In order to make the experience consistent for other developers, you should use the loader-utils to get the loader options:
+### Stateless
 
-```javascript
-const loaderUtils = require("loader-utils");
+Make sure the loader does not retain state between module transformations. Each run should always be independent of other compiled modules as well as previous compilations of the same module.
+
+### Loader Utilities
+
+Take advantage of the [`loader-utils`](https://github.com/webpack/loader-utils) package. It provides a variety of useful tools but one of the most common is retrieving the options passed to the loader:
+
+``` js
+const loaderUtils = require('loader-utils');
 
 module.exports = function(source) {
-    const options = loaderUtils.getOptions(this);
+  const options = loaderUtils.getOptions(this);
 };
 ```
 
-There are also other utility functions like `interpolateName`.
+### Loader Dependencies
 
-### Mark dependencies
+If a loader uses external resources (i.e. by reading from filesystem), they __must__ indicate it. This information is used to invalidate cacheable loaders and recompile in watch mode. Here's a brief example of how to accomplish this using the `addDependency` method:
 
-If a loader uses external resources (i. e. by reading from filesystem), they **must** tell about that. This information is used to invalidate cacheable loaders and recompile in watch mode.
+``` js
+const path = require('path');
 
-``` javascript
-// Loader adding a header
-var path = require("path");
 module.exports = function(source) {
```

Comment:
```js
import { getOptions } from 'loader-utils'
import { validateOptions } from 'schema-utils'
import schema from './options.json' // || schema.json

export default function loader (src, map) {
    const options = getOptions(this)

    validateOptions(schema, options, `${name} Loader`)

   // Code here... e.g
   src = transform(src, options)

   return `export default ${JSON.stringify(src)}`
}
```

Using ES2015 Module Syntax (`import/export`) is fine for loaders when using `webpack >= v2.0.0`, since webpack [resolves](https://github.com/webpack/loader-runner/blob/master/lib/loadLoader.js#L31) them

Reply:
Ah cool, didn't know that.